Add config flow for datadog

avedor opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Implement config flow for the datadog integration and updates the code to meet Bronze standards

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/39839
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

MartinHjelmare commented on script/hassfest/quality_scale.py:
```diff
@@ -269,7 +269,6 @@ class Rule:
     "currencylayer",
     "daikin",
     "danfoss_air",
-    "datadog",
```
Please update the quality scale in a separate PR.

avedor replied:
Sorry I had to rewrite history, but I'll add the quality scale in a follow up!

avedor committed:
Add config flow for datadog

frenck commented on homeassistant/components/datadog/__init__.py:
```diff
@@ -1,65 +1,50 @@
-DEFAULT_RATE = 1
-DOMAIN = "datadog"
-
-CONFIG_SCHEMA = vol.Schema(
-    {
-        DOMAIN: vol.Schema(
-            {
-                vol.Required(CONF_HOST, default=DEFAULT_HOST): cv.string,
-                vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
-                vol.Optional(CONF_PREFIX, default=DEFAULT_PREFIX): cv.string,
-                vol.Optional(CONF_RATE, default=DEFAULT_RATE): vol.All(
-                    vol.Coerce(int), vol.Range(min=1)
-                ),
-            }
-        )
-    },
-    extra=vol.ALLOW_EXTRA,
-)
```
This is still needed in order to creating an migration/import for existing users, right? We need to ensure what we import matches the expected data.

avedor replied:
Readded

frenck commented on homeassistant/components/datadog/__init__.py:
```diff
@@ -1,65 +1,50 @@
+type DatadogConfigEntry = ConfigEntry[DogStatsd]
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up the Datadog component from YAML config."""
+    return True
```
Shouldn't the import be triggered here?

avedor replied:
I think i've set this up correctly now?

frenck commented on homeassistant/components/datadog/__init__.py:
```diff
@@ -1,65 +1,50 @@
+    """Set up Datadog from a config entry."""
 
+    conf = entry.data
+    host = conf.get("host", DEFAULT_HOST)
+    port = conf.get("port", DEFAULT_PORT)
```
According to the configuration flow schema, these are required values. As you are using .get() here, how could they be missing?

avedor replied:
Extraneous defaults removed!

frenck commented on homeassistant/components/datadog/__init__.py:
```diff
@@ -68,38 +53,51 @@ def logbook_entry_listener(event):
+            pass
 
-        statsd.gauge(metric, value, sample_rate=sample_rate, tags=tags)
+    hass.bus.async_listen(EVENT_LOGBOOK_ENTRY, logbook_entry_listener)
+    hass.bus.async_listen(EVENT_STATE_CHANGED, state_changed_listener)
```
We need to unsubscribe from these events when the integration is unloaded, right?

avedor replied:
I think i'm correctly unsubscribing now by using `entry.async_on_unload()`

frenck commented on homeassistant/components/datadog/__init__.py:
```diff
@@ -68,38 +53,51 @@ def logbook_entry_listener(event):
+
+async def validate_datadog_connection(host: str, port: int, prefix: str) -> bool:
+    """Attempt to send a test metric to the Datadog agent."""
+    statsd_client = DogStatsd(host=host, port=port, namespace=prefix)
```
This creates a new connection again? Why not directly use the one created above?

avedor replied:
Switched to reusing the one created above!

frenck commented on homeassistant/components/datadog/strings.json:
```diff
@@ -0,0 +1,21 @@
+      "user": {
+        "description": "Enter your Datadog agent's address and port.",
+        "data": {
+          "host": "Host",
+          "port": "Port",
```
```suggestion
          "host": "[%key:common::config_flow::data::host%]",
          "port": "[%key:common::config_flow::data::port%]",
```

frenck commented on homeassistant/components/datadog/config_flow.py:
```diff
@@ -0,0 +1,55 @@
+                user_input["prefix"],
+            )
+            if not success:
+                errors["base"] = "cannot_connect"
```
`cannot_connect` is missing in the translation file

avedor replied:
Added!

frenck commented on homeassistant/components/datadog/config_flow.py:
```diff
@@ -0,0 +1,55 @@
+            if not success:
+                errors["base"] = "cannot_connect"
+            else:
+                await self.async_set_unique_id(
+                    f"{user_input['host']}:{user_input['port']}"
+                )
```
Host and port do not qualify for a unique identifier, as they can change.

avedor replied:
Removed the unique id declaration

frenck reviewed (dismissed):
Oh hi there @avedor 👋 

Some initial review comments. However the most important review comment:

- Include tests! We require tests for the configuration flows and they need to have 100% coverage.
- Also the YAML import/migration needs to be tested.

../Frenck

avedor committed:
Readd schema; remove defaults; unload listeners

avedor committed:
Add datadog config flow tests

avedor commented:
Hi @frenck!

Super excited to contribute! I think I've addressed all your comments/suggestions and the test suite is reading 100%! If there's anything else that needs changing, please let me know!

gjohansson-ST commented on homeassistant/components/datadog/__init__.py:
```diff
@@ -17,21 +18,25 @@
             {
                 vol.Required(CONF_HOST, default=DEFAULT_HOST): cv.string,
-                vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
+                vol.Optional("port", default=DEFAULT_PORT): cv.port,
```
```suggestion
                vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
```

gjohansson-ST commented on homeassistant/components/datadog/__init__.py:
```diff
@@ -43,63 +48,100 @@
+async def async_setup_entry(hass: HomeAssistant, entry: DatadogConfigEntry) -> bool:
+    """Set up Datadog from a config entry."""
+
+    conf = {**entry.data, **entry.options}
```
```suggestion
    conf = entry.options
```

gjohansson-ST commented on homeassistant/components/datadog/__init__.py:
```diff
@@ -43,63 +48,100 @@
     return True
+
+
+async def validate_datadog_connection(client: DogStatsd) -> bool:
```
Move this function to `config_flow.py` as that seems to be where it's being used.

gjohansson-ST commented on homeassistant/components/datadog/config_flow.py:
```diff
@@ -0,0 +1,85 @@
+            if not success:
+                errors["base"] = "cannot_connect"
+            else:
+                return self.async_create_entry(title="Datadog", data=user_input)
```
```suggestion
                return self.async_create_entry(title="Datadog", data={}, options=user_input)
```
As all these can change, might as well store them in `options` directly and not use `data`

gjohansson-ST commented on homeassistant/components/datadog/config_flow.py:
```diff
@@ -0,0 +1,85 @@
+            step_id="user",
+            data_schema=vol.Schema(
+                {
+                    vol.Required("host", default=DEFAULT_HOST): str,
+                    vol.Required("port", default=DEFAULT_PORT): int,
+                    vol.Required("prefix", default=DEFAULT_PREFIX): str,
+                    vol.Required("rate", default=DEFAULT_RATE): int,
```
```suggestion
                    vol.Required(CONF_HOST, default=DEFAULT_HOST): str,
                    vol.Required(CONF_PORT, default=DEFAULT_PORT): int,
                    vol.Required(CONF_PREFIX, default=DEFAULT_PREFIX): str,
                    vol.Required(CONF_RATE, default=DEFAULT_RATE): int,
```

gjohansson-ST commented on homeassistant/components/datadog/config_flow.py:
```diff
@@ -0,0 +1,85 @@
+    ) -> ConfigFlowResult:
+        """Manage the Datadog options."""
+        if user_input is not None:
+            return self.async_create_entry(title="", data=user_input)
```
```suggestion
            return self.async_create_entry(data=user_input)
```

gjohansson-ST commented on homeassistant/components/datadog/config_flow.py:
```diff
@@ -0,0 +1,85 @@
+        if user_input is not None:
+            return self.async_create_entry(title="", data=user_input)
+
+        data = {**self._config_entry.data, **self._config_entry.options}
```
```suggestion
        data = self._config_entry.options
```

gjohansson-ST commented on homeassistant/components/datadog/strings.json:
```diff
@@ -0,0 +1,24 @@
+{
+  "config": {
```
These strings are missing for options

avedor replied:
Added the options block

gjohansson-ST commented on tests/components/datadog/test_config_flow.py:
```diff
@@ -0,0 +1,150 @@
+    assert result2["data"] == new_config
+
+
+@pytest.mark.asyncio
```
```suggestion
```
All tests are run async so can remove this mark

avedor committed:
Correct variablization; use options

gjohansson-ST commented on homeassistant/components/datadog/const.py:
```diff
@@ -0,0 +1,11 @@
+
+DOMAIN = "datadog"
+
+CONF_PORT = "8125"
```
```suggestion
```

gjohansson-ST commented on homeassistant/components/datadog/config_flow.py:
```diff
@@ -0,0 +1,111 @@
+        )
+
+
+async def validate_datadog_connection(client: DogStatsd) -> bool:
```
```suggestion
async def validate_datadog_connection(hass: HomeAssistant, client: DogStatsd) -> bool:
```

gjohansson-ST commented on homeassistant/components/datadog/config_flow.py:
```diff
@@ -0,0 +1,111 @@
+
+async def validate_datadog_connection(client: DogStatsd) -> bool:
+    """Attempt to send a test metric to the Datadog agent."""
+    loop = asyncio.get_running_loop()
+
+    try:
+        await loop.run_in_executor(None, client.increment, "connection_test")
```
```suggestion
    try:
        hass.async_add_executor_job(client.increment, "connection_test")
```

gjohansson-ST commented on homeassistant/components/datadog/config_flow.py:
```diff
@@ -0,0 +1,111 @@
+class DatadogOptionsFlowHandler(OptionsFlow):
+    """Handle Datadog options."""
+
+    def __init__(self, config_entry: config_entries.ConfigEntry) -> None:
+        """Initialize options flow."""
+        self._config_entry = config_entry
```
```suggestion
```

gjohansson-ST commented on homeassistant/components/datadog/config_flow.py:
```diff
@@ -0,0 +1,111 @@
+        if user_input is not None:
+            return self.async_create_entry(data=user_input)
+
+        data = self._config_entry.options
```
```suggestion
        data = self.config_entry.options
```

gjohansson-ST commented on homeassistant/components/datadog/strings.json:
```diff
@@ -0,0 +1,43 @@
+          "rate": "Rate"
+        },
+        "data_description": {
+          "host": "The hostname or IP address of the Datadog agent.",
+          "port": "Port the Datadog agent is listening on",
+          "prefix": "Metric prefix to use",
+          "rate": "The sample rate of UDP packets sent to Datadog."
```
```suggestion
          "host": "[%key:component::datadog::config::step::user::data_description::host%]",
          "port": "[%key:component::datadog::config::step::user::data_description::port%]",
          "prefix": "[%key:component::datadog::config::step::user::data_description::prefix%]",
          "rate": "[%key:component::datadog::config::step::user::data_description::rate%]"
```

gjohansson-ST commented on homeassistant/components/datadog/strings.json:
```diff
@@ -0,0 +1,43 @@
+        "data": {
+          "host": "[%key:common::config_flow::data::host%]",
+          "port": "[%key:common::config_flow::data::port%]",
+          "prefix": "Prefix",
+          "rate": "Rate"
```
```suggestion
          "prefix": "[%key:component::datadog::config::step::user::data::prefix%]",
          "rate": "[%key:component::datadog::config::step::user::data::rate%]"
```

gjohansson-ST commented on homeassistant/components/datadog/__init__.py:
```diff
@@ -43,63 +48,84 @@
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN,
+            context={"source": "import"},
```
```suggestion
            context={"source": SOURCE_IMPORT},
```
`from homeassistant.config_entries import SOURCE_IMPORT`

gjohansson-ST commented on tests/components/datadog/test_config_flow.py:
```diff
@@ -0,0 +1,144 @@
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"], user_input=MOCK_CONFIG
+        )
+        assert result2["type"] == FlowResultType.FORM
```
Tests should end with `CREATE_ENTRY` or `ABORT`.
Please continue this test with a successful connection so it ends with `CREATE_ENTRY`

gjohansson-ST commented on tests/components/datadog/test_config_flow.py:
```diff
@@ -0,0 +1,144 @@
+    assert result2["data"] == new_config
+
+
+async def test_validate_connection_success() -> None:
```
Move to `test_init.py` as this and tests below are not related to the config flow.

gjohansson-ST commented on tests/components/datadog/test_init.py:
```diff
@@ -145,3 +153,17 @@ async def test_state_changed(hass: HomeAssistant) -> None:
             assert not mock_statsd.gauge.called
+
+
+async def test_unload_entry(hass: HomeAssistant) -> None:
```
This test should load the integration instead with mocking the lib and assert it from there.
We don't want to use the internals directly in tests.

avedor committed:
Correct tests; dry strings

avedor commented:
hey @gjohansson-ST! thank you for all the comments/suggestions! it's my first contribution and i didn't realize how much i could extend from elsewhere! i think i've addressed everything, but if there are any other changes that are needed, i'm happy to do it

gjohansson-ST commented:
> hey @gjohansson-ST! thank you for all the comments/suggestions! it's my first contribution and i didn't realize how much i could extend from elsewhere! i think i've addressed everything, but if there are any other changes that are needed, i'm happy to do it

As a first time contributor of course there's plenty of unknown things. I'll give another pass later and maybe that's it then. 
Btw; I would appreciate if you "resolve" the comments as you fix them to not leave them open on the PR (unless something needs to be discussed). 👍

avedor commented:
> > hey @gjohansson-ST! thank you for all the comments/suggestions! it's my first contribution and i didn't realize how much i could extend from elsewhere! i think i've addressed everything, but if there are any other changes that are needed, i'm happy to do it
> 
> As a first time contributor of course there's plenty of unknown things. I'll give another pass later and maybe that's it then. Btw; I would appreciate if you "resolve" the comments as you fix them to not leave them open on the PR (unless something needs to be discussed). 👍

Thank you so much!

Ah that's a habit from work to leave them open as it's the commenter there who needs to resolve it. I will get on it!

avedor committed:
Fix tests

gjohansson-ST commented on homeassistant/components/datadog/config_flow.py:
```diff
@@ -0,0 +1,101 @@
+                errors["base"] = "cannot_connect"
+            else:
+                return self.async_create_entry(
+                    title="Datadog", data={}, options=user_input
```
```suggestion
                    title=f"Datadog {user_input["host"]}", data={}, options=user_input
```
Since you can have many maybe append the host to the config entry title.

gjohansson-ST commented on homeassistant/components/datadog/config_flow.py:
```diff
@@ -0,0 +1,101 @@
+            # Validate connection to Datadog Agent
+            success = await validate_datadog_connection(
+                self.hass,
+                DogStatsd(user_input["host"], user_input["port"], user_input["prefix"]),
```
```suggestion
                user_input,
```

gjohansson-ST commented on homeassistant/components/datadog/config_flow.py:
```diff
@@ -0,0 +1,101 @@
+async def validate_datadog_connection(hass: HomeAssistant, client: DogStatsd) -> bool:
+    """Attempt to send a test metric to the Datadog agent."""
+    try:
+        hass.async_add_executor_job(client.increment, "connection_test")
```
```suggestion
        DogStatsd(user_input["host"], user_input["port"], user_input["prefix"]),
        hass.async_add_executor_job(client.increment, "connection_test")
```

gjohansson-ST commented on homeassistant/components/datadog/config_flow.py:
```diff
@@ -0,0 +1,101 @@
+        )
+
+
+async def validate_datadog_connection(hass: HomeAssistant, client: DogStatsd) -> bool:
```
```suggestion
async def validate_datadog_connection(hass: HomeAssistant, user_input: dict[str, Any]) -> bool:
```

gjohansson-ST commented on homeassistant/components/datadog/config_flow.py:
```diff
@@ -0,0 +1,101 @@
+    ) -> ConfigFlowResult:
+        """Manage the Datadog options."""
+        if user_input is not None:
+            return self.async_create_entry(data=user_input)
```
I think we should use `validate_datadog_connection()` here to and validate the connection before saving.

gjohansson-ST commented on tests/components/datadog/test_config_flow.py:
```diff
@@ -0,0 +1,126 @@
+async def test_user_flow_success(hass: HomeAssistant) -> None:
+    """Test user-initiated config flow."""
+    with patch(
+        "homeassistant.components.datadog.config_flow.validate_datadog_connection",
```
```suggestion
        "homeassistant.components.datadog.config_flow.DogStatsd.increment",
```
We should mock the library and not the internal function in `config_flow.py`

gjohansson-ST commented on tests/components/datadog/test_config_flow.py:
```diff
@@ -0,0 +1,126 @@
+            }
+        )
+        assert result["type"] == FlowResultType.FORM
+        assert "cannot_connect" in result["errors"].get("base", "")
```
```suggestion
        assert result["errors"] = {"base":"cannot_connect"}
```

avedor committed:
Validate connection before saving; remove socket patching

avedor commented on tests/components/datadog/test_init.py:
```diff
@@ -25,7 +28,8 @@ async def test_datadog_setup_full(hass: HomeAssistant) -> None:
         patch("homeassistant.components.datadog.initialize") as mock_init,
-        patch("homeassistant.components.datadog.statsd"),
+        patch("homeassistant.components.datadog.DogStatsd"),
+        patch("socket.socket", autospec=True),
```
patching `datadog.DogStatsd` was still calling out to socket so i patched it. switching to `datadog.config_flow.DogStatsd` prevented that call and let me remove the socket patch

joostlek commented on homeassistant/components/datadog/__init__.py:
```diff
@@ -43,63 +49,84 @@
-    sample_rate = conf[CONF_RATE]
-    prefix = conf[CONF_PREFIX]
+    conf = entry.options
+    host = conf["host"]
+    port = conf["port"]
+    prefix = conf["prefix"]
+    sample_rate = conf["rate"]
```
can we use constants for this?

avedor replied:
i added them like this:

```
    conf = entry.options
    host = conf[CONF_HOST]
    port = conf[CONF_PORT]
    prefix = conf[CONF_PREFIX]
    sample_rate = conf[CONF_RATE]
```

or would it be better to have them as globals?

avedor replied:
awesome!

joostlek commented on homeassistant/components/datadog/__init__.py:
```diff
@@ -43,63 +49,84 @@
-    port = conf[CONF_PORT]
-    sample_rate = conf[CONF_RATE]
-    prefix = conf[CONF_PREFIX]
+    conf = entry.options
```
why is this stored in the entry options? Things like host and port sound like the bare minimum to connect, and thus belong in data

joostlek replied:
Let me check with him then

avedor replied:
i just appreciate the time and detail you both are giving this!

i think i've gotten it split everywhere and the tests are passing, so i'm hopeful that looks good! i'll add a `reconfigure` PR to my list and plan to tackle this once this PR is done!

joostlek commented on homeassistant/components/datadog/config_flow.py:
```diff
@@ -0,0 +1,113 @@
+)
+
+
+class DatadogConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
```
```suggestion
class DatadogConfigFlow(ConfigFlow, domain=DOMAIN):
```

joostlek commented on homeassistant/components/datadog/config_flow.py:
```diff
@@ -0,0 +1,113 @@
+            errors=errors,
+        )
+
+    async def async_step_import(self, import_config: dict) -> ConfigFlowResult:
+        """Handle import from configuration.yaml."""
+        return await self.async_step_user(import_config)
```
The import flow should not return errors. Instead it should abort. We should also create a repair issue explaining the user that they should remove the YAML config, or, if it couldn't connect, tell them they couldnt import because we couldnt commit.

I don't think there are many integrations which have it implemented right now, so be sure to check for merged PRs with the config-flow label

avedor committed:
Add repair issue

avedor committed:
Correct tests for yaml issue

avedor committed:
Split data and options

avedor committed:
Move issues to new file; update tests

avedor committed:
Correct import issue tests

avedor commented on homeassistant/components/datadog/strings.json:
```diff
@@ -0,0 +1,59 @@
+      "cannot_connect": "[%key:common::config_flow::error::cannot_connect%]"
+    },
+    "abort": {
+      "already_configured_service": "[%key:common::config_flow::abort::already_configured_service%]"
```
i realize it's theoretically possible that someone would want to send to multiple datadog agents, so could have multiple configs

gjohansson-ST replied:
I meant that the strings need to match, not that it should be removed.
It's a good check as it doesn't make sense to have two config entries for the same host:port combination

avedor replied:
Ah yeah, I can see why having a duplication check for the specific host:port combo would be helpful! added it back in and got test coverage for it

avedor committed:
Check duplicate configs

gjohansson-ST commented on homeassistant/components/datadog/issues.py:
```diff
@@ -0,0 +1,42 @@
+    if import_success:
+        async_create_issue(
+            hass,
+            DOMAIN,
```
Use HOMEASSISTANT_DOMAIN instead so it uses the already existing strings there instead.

gjohansson-ST commented on homeassistant/components/datadog/strings.json:
```diff
@@ -0,0 +1,59 @@
+    }
+  },
+  "issues": {
+    "deprecated_yaml": {
+      "title": "{domain} YAML configuration is deprecated",
+      "description": "Configuring {domain} using YAML is being removed. Please remove the {domain} YAML configuration from your configuration.yaml file and continue to set up the integration manually."
+    },
```
```suggestion
```

gjohansson-ST commented on tests/components/datadog/test_init.py:
```diff
@@ -145,3 +166,97 @@ async def test_state_changed(hass: HomeAssistant) -> None:
+        assert not mock_dogstatsd.gauge.called
+
+
+async def test_yaml_import_creates_deprecation_issue(hass: HomeAssistant) -> None:
```
We already have these test cases in `test_config_flow.py` so extend it there to assert issues has been created. No need to have separate test cases.

avedor committed:
Combine tests

avedor commented on homeassistant/components/datadog/config_flow.py:
```diff
@@ -0,0 +1,148 @@
+            success = await validate_datadog_connection(
+                self.hass,
+                user_input,
+            )
```
i had to do a little extra logic to build the config, so have this now as:

```
        connection_data = {
            CONF_HOST: data[CONF_HOST],
            CONF_PORT: data[CONF_PORT],
            CONF_PREFIX: user_input.get(CONF_PREFIX, options[CONF_PREFIX]),
            CONF_RATE: user_input.get(CONF_RATE, options[CONF_RATE]),
        }

        success = await validate_datadog_connection(
            self.hass,
            connection_data,
        )
```

avedor replied:
nevermind! got the tests all set!

avedor committed:
Add additional dupe check

avedor committed:
Fixup connection params

gjohansson-ST commented on homeassistant/components/datadog/strings.json:
```diff
@@ -0,0 +1,55 @@
+      "cannot_connect": "[%key:common::config_flow::error::cannot_connect%]"
+    },
+    "abort": {
+      "already_configured": "[%key:common::config_flow::abort::already_configured_service%]"
```
We need an abort `cannot_connect` as well (as you have that as an abort reason when import fails on connection issues.)

gjohansson-ST commented on tests/components/datadog/test_config_flow.py:
```diff
@@ -0,0 +1,219 @@
+        )
+
+        assert result2["type"] == FlowResultType.FORM
+        assert result2["errors"] == {"base": "cannot_connect"}
```
Please continue this test case so it results in `CREATE_ENTRY`

avedor replied:
i extended it the same way i did for `test_user_flow_retry_after_connection_fail`. i think that's right, but not positive

gjohansson-ST commented on tests/components/datadog/test_config_flow.py:
```diff
@@ -0,0 +1,219 @@
+        assert result2["errors"] == {"base": "cannot_connect"}
+
+
+async def test_import_flow(hass: HomeAssistant) -> None:
```
```suggestion
async def test_import_flow(hass: HomeAssistant, issue_registry: ir.IssueRegistry) -> None:
```

gjohansson-ST commented on tests/components/datadog/test_config_flow.py:
```diff
@@ -0,0 +1,219 @@
+
+        await hass.async_block_till_done()
+
+        # Deprecation issue should be created
+        issue_registry = ir.async_get(hass)
+        issue = issue_registry.issues.get(
+            ("homeassistant", f"deprecated_yaml_{datadog.DOMAIN}")
+        )
```
```suggestion
        # Deprecation issue should be created
        issue = issue_registry.async_get_issue(
            HOMEASSISTANT_DOMAIN, "deprecated_yaml_datadog"
        )
```

gjohansson-ST commented on tests/components/datadog/test_config_flow.py:
```diff
@@ -0,0 +1,219 @@
+        assert issue.translation_key == "deprecated_yaml"
+        assert issue.severity == ir.IssueSeverity.WARNING
+
+    # Invalid import with connection error
```
Put this in a separate test for connection error from import

gjohansson-ST commented on tests/components/datadog/test_init.py:
```diff
@@ -51,20 +56,31 @@ async def test_datadog_setup_defaults(hass: HomeAssistant) -> None:
             datadog.DOMAIN,
-            {datadog.DOMAIN: {"host": "host", "rate": datadog.DEFAULT_RATE}},
+            {
+                datadog.DOMAIN: {
```
In a a follow-up PR it would be good to change all these tests to load from config entry instead so a clean-up of the yaml config is easier when it's time for that.

gjohansson-ST replied:
Yeah, rather easy to just make a config entry and load that instead of the yaml config.
but let's do that in a follow-up

gjohansson-ST reviewed (changes requested):
Only very small things left now it seems. Looks good 👍

avedor committed:
Fix strings; separate tests

avedor committed:
Replace issue registry call

avedor committed:
Remove extraneous parens

gjohansson-ST commented on homeassistant/components/datadog/strings.json:
```diff
@@ -0,0 +1,56 @@
+    "error": {
+      "cannot_connect": "[%key:common::config_flow::error::cannot_connect%]"
+    },
+    "abort": {
```
These abort messages also need to exist in `options`.

gjohansson-ST commented on homeassistant/components/datadog/strings.json:
```diff
@@ -0,0 +1,56 @@
+      "init": {
+        "description": "Update the Datadog configuration.",
+        "data": {
+          "host": "[%key:common::config_flow::data::host%]",
+          "port": "[%key:common::config_flow::data::port%]",
```
```suggestion
```

gjohansson-ST commented on homeassistant/components/datadog/strings.json:
```diff
@@ -0,0 +1,56 @@
+          "rate": "[%key:component::datadog::config::step::user::data::rate%]"
+        },
+        "data_description": {
+          "host": "[%key:component::datadog::config::step::user::data_description::host%]",
+          "port": "[%key:component::datadog::config::step::user::data_description::port%]",
```
```suggestion
```

gjohansson-ST commented on homeassistant/components/datadog/strings.json:
```diff
@@ -0,0 +1,56 @@
+  "issues": {
+    "deprecated_yaml_import_connection_error": {
+      "title": "{domain} YAML configuration import failed",
+      "description": "[%key:component::datadog::config::abort::cannot_connect%]"
```
```suggestion
      "description": "There was an error connecting to the Datadog Agent when trying to import the YAML configuration.\n\nEnsure the YAML configuration is correct and restart Home Assistant to try again or remove the {domain} configuration from your `configuration.yaml` file and continue to [set up the integration]({url}) manually."
```

gjohansson-ST commented on homeassistant/components/datadog/strings.json:
```diff
@@ -0,0 +1,56 @@
+    },
+    "abort": {
+      "already_configured": "[%key:common::config_flow::abort::already_configured_service%]",
+      "cannot_connect": "There was an error connecting to the Datadog Agent when trying to import the YAML configuration.\n\nEnsure the YAML configuration is correct and restart Home Assistant to try again or remove the {domain} configuration from your `configuration.yaml` file and continue to [set up the integration]({url}) manually."
```
```suggestion
      "cannot_connect": "[%key:common::config_flow::error::cannot_connect%]"
```

gjohansson-ST commented on tests/components/datadog/test_config_flow.py:
```diff
@@ -0,0 +1,233 @@
+        domain=datadog.DOMAIN,
+        data=MOCK_DATA,
+        options=MOCK_OPTIONS,
+        unique_id="datadog_unique",
```
```suggestion
```
Integration does not have unique id

gjohansson-ST commented on tests/components/datadog/test_config_flow.py:
```diff
@@ -0,0 +1,233 @@
+    hass: HomeAssistant, issue_registry: ir.IssueRegistry
+) -> None:
+    """Test import triggers config flow and is accepted."""
+    # Success case: Valid import
```
```suggestion
```

gjohansson-ST commented on tests/components/datadog/test_config_flow.py:
```diff
@@ -0,0 +1,233 @@
+    hass: HomeAssistant, issue_registry: ir.IssueRegistry
+) -> None:
+    """Test import triggers connection error issue."""
+    # Invalid import with connection error
```
```suggestion
```

gjohansson-ST commented on tests/components/datadog/test_config_flow.py:
```diff
@@ -0,0 +1,233 @@
+        domain=datadog.DOMAIN,
+        data=MOCK_DATA,
+        options=MOCK_OPTIONS,
+        unique_id="datadog_unique",
```
```suggestion
```

gjohansson-ST commented on tests/components/datadog/test_init.py:
```diff
@@ -145,3 +158,53 @@ async def test_state_changed(hass: HomeAssistant) -> None:
+        )
+        entry.add_to_hass(hass)
+
+        assert await hass.config_entries.async_setup(entry.entry_id)
+        await hass.async_block_till_done()
+
+        assert await hass.config_entries.async_unload(entry.entry_id)
+        await hass.async_block_till_done()
```
```suggestion
        assert await hass.config_entries.async_setup(entry.entry_id)
        await hass.async_block_till_done()
        
        assert entry.state is ConfigEntryState.LOADED

        assert await hass.config_entries.async_unload(entry.entry_id)
        await hass.async_block_till_done()
        
        assert entry.state is ConfigEntryState.NOT_LOADED
```

avedor committed:
Fixup strings

avedor committed:
Fix more strings

gjohansson-ST committed:
Merge branch 'dev' into feat/datadog-config-flow

gjohansson-ST reviewed (approved):
Thanks @avedor 👍

joostlek commented on tests/components/datadog/test_config_flow.py:
```diff
@@ -0,0 +1,229 @@
+        data=MOCK_CONFIG,
+    )
+
+    assert result["type"] == "abort"
```
We can use FlowResultType here

joostlek commented on tests/components/datadog/test_config_flow.py:
We're missing a test where we test setting up the same entry twice via the user flow

avedor replied:
I know there a couple things that need follow up PRs already, I will add this and the your comment re FlowResultType to that list!

gjohansson-ST replied:
You can ping me in those PR's directly so we can have it sorted before beta comes if you have the time for it.

avedor commented:
Thank you all for your patience and for your hard work! I really appreciate all of your guidance on this!
